src/hotspot/share/opto/regmask.hpp

```diff
@@ -260,115 +548,299 @@ class RegMask {
   static int num_registers(uint ireg);
   static int num_registers(uint ireg, LRG &lrg);
 
-  // Fast overlap test.  Non-zero if any registers in common.
+  // Overlap test. Non-zero if any registers in common, including all-stack.
   bool overlap(const RegMask &rm) const {
+    assert(_offset == rm._offset, "offset mismatch");
     assert(valid_watermarks() && rm.valid_watermarks(), "sanity");
+
+    // Very common overlap case: _rm_up overlap. Check first to reduce
+    // execution time.
     unsigned hwm = MIN2(_hwm, rm._hwm);
     unsigned lwm = MAX2(_lwm, rm._lwm);
-    uintptr_t result = 0;
     for (unsigned i = lwm; i <= hwm; i++) {
-      result |= _RM_UP[i] & rm._RM_UP[i];
+      if (_rm_up(i) & rm._rm_up(i)) {
+        return true;
+      }
+    }
+
+    // Very rare overlap cases below.
+
+    // We are both all-stack
+    if (is_AllStack() && rm.is_AllStack()) {
+      return true;
     }
-    return result;
+
+    // We are all-stack and rm _hwm is bigger than us
+    if (is_AllStack() && rm._hwm >= _rm_size) {
+      for (unsigned i = MAX2(rm._lwm, _rm_size); i <= rm._hwm; i++) {
+        if (rm._rm_up(i)) {
+          return true;
+        }
+      }
+    }
+
+    // rm is all-stack and our _hwm is bigger than rm
+    if (rm.is_AllStack() && _hwm >= rm._rm_size) {
+      for (unsigned i = MAX2(_lwm, rm._rm_size); i <= _hwm; i++) {
+        if (_rm_up(i)) {
+          return true;
+        }
+      }
+    }
+
+    // No overlap (also very common)
+    return false;
   }
 
   // Special test for register pressure based splitting
   // UP means register only, Register plus stack, or stack only is DOWN
   bool is_UP() const;
 
-  // Clear a register mask
+  // Clear a register mask. Does not clear any offset.
   void Clear() {
-    _lwm = _RM_MAX;
+    _lwm = _rm_max();
     _hwm = 0;
-    memset(_RM_UP, 0, sizeof(uintptr_t) * _RM_SIZE);
+    _set_range(0, 0, _rm_size);
+    set_AllStack(false);
     assert(valid_watermarks(), "sanity");
   }
 
   // Fill a register mask with 1's
   void Set_All() {
+    assert(_offset == 0, "offset non-zero");
+    Set_All_From_Offset();
+  }
+
+  // Fill a register mask with 1's from the current offset.
+  void Set_All_From_Offset() {
     _lwm = 0;
-    _hwm = _RM_MAX;
-    memset(_RM_UP, 0xFF, sizeof(uintptr_t) * _RM_SIZE);
+    _hwm = _rm_max();
+    _set_range(0, 0xFF, _rm_size);
+    set_AllStack(true);
```

Comment:
```suggestion
    set_AllStack();
```
You already have a default `true` value, right? Check for other occurances.

Reply:
Thanks, I just removed the default value and made it explicit everywhere.